### PR TITLE
Add rule for space around parenthesis

### LINKED ISSUE #3889: Whitespace around parenthesis
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

It seems there is no rule that flags issue around missing whitespaces around keywords, or missing whitespaces around parenthesis.

For example,  the following query is considered valid (and it is, syntactically).
```
SELECT * FROM(SELECT 1 AS C1)AS T1;
```
But  it would be great if sqlfluff can point at missing whitespace before opening parenthesis and after closing parenthesis (or, to put it in another way - whitespace before and after a keyword).

e.g. A fix for the above query will be:
```
SELECT * FROM (SELECT 1 AS C1) AS T1;
```

### Use case

When brackets/parenthesis are used with sql keywords, the query in most of the dialects (if not all) is valid, but is not very readable. A rule to flag such issues would be great.

### Dialect

All

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by NelsonTorres**
### Brief summary of the change made

fixes #3889

### Are there any other side effects of this change that we should be aware of?
I don't think so

### Pull Request checklist
- [ X] Please confirm you have completed any of the necessary steps below.

I didn't add test cases at " `.sql`/`.yml` parser test cases in `test/fixtures/dialects` " can anyone confirm if it is needed since I have only added a new rule?

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by alanmcruickshank** [comment]
This looks like great usage of the new reflow code! I'm AFK for the next week so can't review this properly yet, but this is basically what I was hoping you would do 👍

**Comment by barrywhart** [comment]
Have you looked at the failing tests? A Snowflake dialect test is failing.

**Comment by NelsonTorres** [comment]
The tests are failing when parsing functions, a simple carbon copy will not work out.

e.g:  
`   select foo(bar) from T1; `
Will ident to
`   select foo (bar) from T1;`

Do you have any recomendation on how to fix this?

**Comment by barrywhart** [comment]
What do you mean here?
>a simple carbon copy will not work out.

**Comment by NelsonTorres** [comment]
> 
You can ignore my prior comment, 
I was failing to see that by using provide_raw_stack=True I could look backwards and fix my error

**Comment by NelsonTorres** [comment]
@alanmcruickshank  I believe all the tests are fixed now, can you approve the workflows again?

**Comment by barrywhart** [comment]
@NelsonTorres: I'm seeing some failures. I haven't had time to look at the details.

File: src/sqlfluff/rules/L071.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        SELECT * FROM (SELECT 1 AS C1) AS T1;
+
+    """
+
+    groups = ("all", "core")
+    crawl_behaviour = SegmentSeekerCrawler(
+        {"start_bracket", "end_bracket"}, provide_raw_stack=True
+    )
+
+    def _eval(self, context: RuleContext) -> List[LintResult] | None:
```
Comment (alanmcruickshank) [inline src/sqlfluff/rules/L071.py:42]: the syntax `List[LintResult] | None` is only supported from python 3.10 onwards. Can you use `Optional[List[LintResult]]` instead. That should solve some of your failing test cases.

**Review by alanmcruickshank — CHANGES_REQUESTED** [review]
Tests are failing, but I think I know why. If you resolve that I'll do another review again once we know all the tests pass.

File: src/sqlfluff/rules/L071.py (hunk 1 above)
Comment (NelsonTorres) [reply src/sqlfluff/rules/L071.py]: This issue is now fixed with my latest commit

**Review by alanmcruickshank — APPROVED** [review]
Looks good to me!

**Pull request merged**